group-coordinator/src/test/java/org/apache/kafka/coordinator/group/consumer/TargetAssignmentBuilderTest.java

```diff
@@ -0,0 +1,698 @@
+                    member,
+                    targetAssignments.getOrDefault(memberId, Assignment.EMPTY),
+                    subscriptionMetadata
+                ));
+            });
+
+            // All the updated are added and all the deleted
+            // members are removed.
+            updatedMembers.forEach((memberId, updatedMemberOrNull) -> {
+                if (updatedMemberOrNull == null) {
+                    memberSpecs.remove(memberId);
+                } else {
+                    memberSpecs.put(memberId, createAssignmentMemberSpec(
+                        updatedMemberOrNull,
+                        targetAssignments.getOrDefault(memberId, Assignment.EMPTY),
+                        subscriptionMetadata
+                    ));
+                }
+            });
+
+            // Prepare the expected topic metadata.
+            Map<Uuid, AssignmentTopicMetadata> topicMetadata = new HashMap<>();
+            subscriptionMetadata.forEach((topicName, metadata) -> {
+                topicMetadata.put(metadata.id(), new AssignmentTopicMetadata(metadata.numPartitions()));
+            });
+
+            // Prepare the expected assignment spec.
+            AssignmentSpec assignmentSpec = new AssignmentSpec(
+                memberSpecs,
+                topicMetadata
+            );
+
+            // We use `any` here to always return an assignment but use `verify` later on
+            // to ensure that the input was correct.
+            when(assignor.assign(any())).thenReturn(new GroupAssignment(assignments));
+
+            // Create and populate the assignment builder.
+            TargetAssignmentBuilder builder = new TargetAssignmentBuilder(groupId, groupEpoch, assignor)
+                .withMembers(members)
+                .withSubscriptionMetadata(subscriptionMetadata)
+                .withTargetAssignments(targetAssignments);
+
+            // Add the updated members or delete the deleted members.
+            updatedMembers.forEach((memberId, updatedMemberOrNull) -> {
+                if (updatedMemberOrNull != null) {
+                    builder.addOrUpdateMember(memberId, updatedMemberOrNull);
+                } else {
+                    builder.removeMember(memberId);
+                }
+            });
+
+            // Execute the builder.
+            TargetAssignmentBuilder.TargetAssignmentResult result = builder.build();
+
+            // Verify that the assignor was called once with the expected
+            // assignment spec.
+            verify(assignor, times(1)).assign(assignmentSpec);
+
+            return result;
+        }
+    }
+
+    @Test
+    public void testCreateAssignmentMemberSpec() {
+        Uuid fooTopicId = Uuid.randomUuid();
+        Uuid barTopicId = Uuid.randomUuid();
+
+        ConsumerGroupMember member = new ConsumerGroupMember.Builder("member-id")
+            .setSubscribedTopicNames(Arrays.asList("foo", "bar", "zar"))
```

Comment:
Is this an example of a subscribed topic where we don't have metadata, so we don't subscribe?

Reply:
Correct. You can see at L245 that the assignor only gets the ids of foo and bar because zar does not exist.